presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java

```diff
@@ -148,6 +149,7 @@
 
     private final AtomicReference<Set<Input>> inputs = new AtomicReference<>(ImmutableSet.of());
     private final AtomicReference<Optional<Output>> output = new AtomicReference<>(Optional.empty());
+    private final AtomicReference<List<TableInfo>> referencedTables = new AtomicReference<>(ImmutableList.of());
```

Comment:
use `null` initial values, to make sure it was not yet set, then in `setReferencedTables` please verify that previous value was `null`

Reply:
Actually, that doesn't work. There are many paths that don't even call the setInput/setOutput/setReferencedTables methods.